bpf: Simplify connection tracking logic

tgraf opened the pull request:
Connection tracking entries get created for non-SYN packets. This allows a persistent connection to be temporarily be removed and then get re-added.

RST no longer triggers immediate deletion of the entry. Instead, directional closure is triggered. This will result in the entry getting evicted if the other direction confirms the RST/FIN or the entry gets reset on the next non RST packet in the direction of the initial RST.

Presence of the connection tracking entry is only used for REPLY and RELATED, otherwise the policy table is always considered for policy purposes. This will guarantee that connections get dropped as soon as policy no longer allows the connection. 

Fixes: #2697 
Fixes: #2056

tgraf committed:
bpf: Simplify connection tracking logic

Connection tracking entries get created for non-SYN packets. This allows a
persistent connection to be temporarily be removed and then get re-added.

RST no longer triggers immediate deletion of the entry. Instead, directional
closure is triggered. This will result in the entry getting evicted if the
other direction confirms the RST/FIN or the entry gets reset on the next non
RST packet in the direction of the initial RST.

Presence of the connection tracking entry is only used for REPLY and RELATED,
otherwise the policy table is always considered for policy purposes. This will
guarantee that connections get dropped as soon as policy no longer allows the
connection.

Fixes: #2697

Signed-off-by: Thomas Graf <thomas@cilium.io>

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -929,18 +928,19 @@ static inline int __inline__ ipv4_policy(struct __sk_buff *skb, int ifindex, __u
+	if (ret != CT_REPLY && ret != CT_RELATED && verdict != TC_ACT_OK)
+		return DROP_POLICY;
+
+
```
whitespace

tgraf replied:
fixed

joestringer commented on bpf/lib/conntrack.h:
```diff
@@ -230,16 +225,10 @@ static inline int __inline__ ct_lookup6(void *map, struct ipv6_ct_tuple *tuple,
 
-			if (unlikely(flags.syn && !flags.ack))
+			if (unlikely(flags.rst || flags.fin))
+				action = ACTION_CLOSE;
```
Ah, nice. My only feedback was that I figure it shouldn't stick around as long as an established connection if it's closed. This will now be a 10s timeout from the time of FIN/RST. Should mean it's cleaned up within 20s. (GC just triggered, next GC in 10s will not clean up the entry, but the following will).

jrfastab replied:
Not required or possible in this patch but in theory hooking into sock_ops you could delete the entry on sock state. Obviously future work though.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -829,17 +829,16 @@ static inline int __inline__ ipv6_policy(struct __sk_buff *skb, int ifindex, __u
 
+	/* Reply packets and related packets are allowed, all others must be
+	 * permitted by policy */
+	if (ret != CT_REPLY && ret != CT_RELATED && verdict != TC_ACT_OK)
```
Interesting subtle change here that means that we no longer rely on CT for established connections; such packets now rely on policy. Makes sense, means we're less reliant on the CT state.

jrfastab replied:
Seems `if ret = CT_ESTABLISHED` the policy_can_access could be skipped and we could jump to the proxy port check directly after ct_lookup(). I'm fine if that goes in as an optimization later but could save some map lookups.

jrfastab replied:
nevermind re-reading code its clear the above comment is not correct. But do I read this correct, if ct_lookup is CT_REPLY or CT_RELATED then we will return 0 sans NAT requirements.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -829,17 +829,16 @@ static inline int __inline__ ipv6_policy(struct __sk_buff *skb, int ifindex, __u
+	if (ret != CT_REPLY && ret != CT_RELATED && verdict != TC_ACT_OK)
+		return DROP_POLICY;
+
+	if (unlikely(ret == CT_NEW)) {
```
by the way, is it really that unlikely to get new connections?

tgraf replied:
Not really, will remove

joestringer reviewed (approved):
LGTM, this is definitely simpler. It's a bit more relaxed (allow ongoing TCP connections we didn't observe the start for or which we timed out because there was no traffic for >5mins), while also being more restrictive (ensure that even for established connections, the policy decision wins).

Just wondering if there might be a side effect that when transitioning from Allow L3/L4 policy -> no (default, allow) policy, maybe the policy maps will be updated before the BPF is regenerated and packets for established connections will be dropped for a short time between updating the maps and generating the BPF with no policy.. but that's a bit of a corner case. +@aanm

tgraf committed:
bpf: Enforce to-world access via CIDR map

Remove ALLOW_TO_WORLD define from the program and enforce all access
to external services via the CIDR map.

This simplifies the code and makes policy updates more atomic.

Signed-off-by: Thomas Graf <thomas@cilium.io>

joestringer reviewed (approved):
New patch LGTM.

aanm commented on pkg/policy/api/rule.go:
```diff
@@ -274,6 +274,9 @@ type EgressRule struct {
 type CIDR string
 
+// CIDRMatchAll is a []CIDR that matches everything
+var CIDRMatchAll = []CIDR{CIDR("0.0.0.0/0"), CIDR("::/0")}
```
Can you change this name? It's not actually matching "All" it's matching all IPs that *don't* belong to the cluster CIDR, right?

tgraf replied:
Yes but that's a datapath specific behaviour. The CIDR itself really matches everything. I know exactly what you mean but we should document the behaviour in one place instead of spreading it out. If we ever change it we have to touch every single place.

aanm replied:
That's true... But this needs to be documented what about in `pkg/policy/rule.go:362`? and `381`?

tgraf replied:
Agreed

aanm commented on pkg/policy/rule.go:
```diff
@@ -355,9 +355,15 @@ func (r *rule) resolveCIDRPolicy(ctx *SearchContext, state *traceState, result *
 
+		for _, fromEntity := range ingressRule.FromEntities {
+			switch fromEntity {
+			case api.EntityWorld:
```
here

aanm commented on pkg/policy/rule.go:
```diff
@@ -367,9 +373,15 @@ func (r *rule) resolveCIDRPolicy(ctx *SearchContext, state *traceState, result *
 
+		for _, toEntity := range egressRule.ToEntities {
+			switch toEntity {
+			case api.EntityWorld:
```
and here

jrfastab commented on bpf/bpf_lxc.c:
```diff
@@ -929,18 +928,18 @@ static inline int __inline__ ipv4_policy(struct __sk_buff *skb, int ifindex, __u
+	if (ret != CT_REPLY && ret != CT_RELATED && verdict != TC_ACT_OK)
+		return DROP_POLICY;
+
+	if (ret == CT_NEW) {
```
Same comment here as above seems we can optimize out the policy_can_access for RELATED flows.

jrfastab reviewed (approved):
Other than a few optimizations and nits LGTM.

tgraf committed:
policy: Document what connections FromCIDR and ToCIDR and apply to

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Split IPv6 handling into separate tail call

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Reduce the number of supported IPv6 extension headers

This greatly reduces the complexity of the program

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Relax verifier in conntrack code

Signed-off-by: Thomas Graf <thomas@cilium.io>
